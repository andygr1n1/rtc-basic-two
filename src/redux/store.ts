import { todoReducer } from '@/components/todo.slice'
import { configureStore } from '@reduxjs/toolkit'

export const rootStore = configureStore({
    reducer: {
        todo$: todoReducer,
    },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type IRootStore = ReturnType<typeof rootStore.getState>

// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type IAppDispatch = typeof rootStore.dispatch
