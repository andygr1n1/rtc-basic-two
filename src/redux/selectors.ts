// reselect -> memoization
import { createSelector } from '@reduxjs/toolkit'
import { IRootStore } from './store'
import { ITodoSlice } from '@/components/todo.slice'
import { ITodo } from '@/interfaces/ITodo.interface'

export const getTodo$: (store: IRootStore) => ITodoSlice = (store) => store.todo$
export const getAllTodos: (store: IRootStore) => ITodo[] = (store) => store.todo$.todos
export const getActiveTodosFilter: (store: IRootStore) => string = (store) => store.todo$.filter

export const getTodosByFilter = createSelector(
    [getAllTodos, getActiveTodosFilter],
    (allTodos, activeFilter): ITodo[] | undefined => {
        if (activeFilter === 'all') return allTodos
        if (activeFilter === 'completed') return allTodos.filter((todo) => todo.completed)
        if (activeFilter === 'active') return allTodos.filter((todo) => !todo.completed)
    },
)

export const TodoListSelector = createSelector([getTodosByFilter, getTodo$], (getTodosByFilter, getTodo$) => {
    return { todos: getTodosByFilter, error: getTodo$.error, loading: getTodo$.loading }
})
