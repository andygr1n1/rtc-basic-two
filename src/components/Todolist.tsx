import { useAppSelector } from '@/redux'
import { TodoListSelector } from '@/redux/selectors'
import { TodoItem } from './TodoItem'

export const Todolist: React.FC = () => {
    const { error, loading, todos } = useAppSelector(TodoListSelector)

    if (error) return <div>it seems to be, something happend on server</div>

    return loading ? (
        <div>loading</div>
    ) : (
        <ul>
            {todos?.map((todo) => (
                <TodoItem key={todo.id} todo={todo} />
            ))}
        </ul>
    )
}
