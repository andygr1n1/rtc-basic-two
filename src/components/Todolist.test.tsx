import { Todolist } from './Todolist'
import { beforeEach, describe, expect, it, afterEach, vitest } from 'vitest'
import { act } from '@testing-library/react'
import ReactDOM from 'react-dom/client'
import * as reduxHooks from '../redux/hooks'

vitest.mock('react-redux')

let container: Element | DocumentFragment

describe('Todolist', () => {
    beforeEach(() => {
        container = document.createElement('div')
        document.body.appendChild(container)
    })
    afterEach(() => {
        container && document.body.removeChild(container)
    })

    it('should create 3 Todos', () => {
        // ;(useAppSelector as Mock).mockReturnValue({ error: true, loading: true, todos: mockTodos })
        vitest.spyOn(reduxHooks, 'useAppSelector').mockReturnValue({ error: false, loading: false, todos: mockTodos })

        act(() => {
            ReactDOM.createRoot(container).render(<Todolist />)
        })

        const TodoList = container.querySelector('ul')
        const Todos = container.querySelectorAll('li')

        expect(TodoList?.tagName).toBe('UL')

        expect(Todos.length).toBe(3)
    })

    it('should create an empty Todolist because of fetch error', () => {
        vitest.spyOn(reduxHooks, 'useAppSelector').mockReturnValue({ error: true, loading: false, todos: mockTodos })

        act(() => {
            ReactDOM.createRoot(container).render(<Todolist />)
        })

        const Todos = container.querySelectorAll('li')

        expect(Todos.length).toBe(0)
    })
})

const mockTodos = [
    {
        userId: 1,
        id: 8,
        title: 'quo adipisci enim quam ut ab',
        completed: true,
    },
    {
        userId: 1,
        id: 9,
        title: 'molestiae perspiciatis ipsa',
        completed: true,
    },
    {
        userId: 1,
        id: 10,
        title: 'illo est ratione doloremque quia maiores aut',
        completed: true,
    },
]
