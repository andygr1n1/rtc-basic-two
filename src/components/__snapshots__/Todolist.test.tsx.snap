// Vitest Snapshot v1

exports[`Todolist > should create an empty Todolist 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul />
    </div>
  </body>,
  "container": <div>
    <ul />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Todolist > should create an empty Todolist 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul>
        <li
          class="flex gap-5"
        >
          <input
            checked=""
            class="peer"
            type="checkbox"
          />
          <div
            class="peer-checked:line-through"
          >
            quo adipisci enim quam ut ab
          </div>
          <span
            class="material-icons-round cursor-pointer hover:!text-red-500"
          >
            delete
          </span>
        </li>
        <li
          class="flex gap-5"
        >
          <input
            checked=""
            class="peer"
            type="checkbox"
          />
          <div
            class="peer-checked:line-through"
          >
            molestiae perspiciatis ipsa
          </div>
          <span
            class="material-icons-round cursor-pointer hover:!text-red-500"
          >
            delete
          </span>
        </li>
        <li
          class="flex gap-5"
        >
          <input
            checked=""
            class="peer"
            type="checkbox"
          />
          <div
            class="peer-checked:line-through"
          >
            illo est ratione doloremque quia maiores aut
          </div>
          <span
            class="material-icons-round cursor-pointer hover:!text-red-500"
          >
            delete
          </span>
        </li>
        <li
          class="flex gap-5"
        >
          <input
            checked=""
            class="peer"
            type="checkbox"
          />
          <div
            class="peer-checked:line-through"
          >
            vero rerum temporibus dolor
          </div>
          <span
            class="material-icons-round cursor-pointer hover:!text-red-500"
          >
            delete
          </span>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul>
      <li
        class="flex gap-5"
      >
        <input
          checked=""
          class="peer"
          type="checkbox"
        />
        <div
          class="peer-checked:line-through"
        >
          quo adipisci enim quam ut ab
        </div>
        <span
          class="material-icons-round cursor-pointer hover:!text-red-500"
        >
          delete
        </span>
      </li>
      <li
        class="flex gap-5"
      >
        <input
          checked=""
          class="peer"
          type="checkbox"
        />
        <div
          class="peer-checked:line-through"
        >
          molestiae perspiciatis ipsa
        </div>
        <span
          class="material-icons-round cursor-pointer hover:!text-red-500"
        >
          delete
        </span>
      </li>
      <li
        class="flex gap-5"
      >
        <input
          checked=""
          class="peer"
          type="checkbox"
        />
        <div
          class="peer-checked:line-through"
        >
          illo est ratione doloremque quia maiores aut
        </div>
        <span
          class="material-icons-round cursor-pointer hover:!text-red-500"
        >
          delete
        </span>
      </li>
      <li
        class="flex gap-5"
      >
        <input
          checked=""
          class="peer"
          type="checkbox"
        />
        <div
          class="peer-checked:line-through"
        >
          vero rerum temporibus dolor
        </div>
        <span
          class="material-icons-round cursor-pointer hover:!text-red-500"
        >
          delete
        </span>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
